File: 496-next-greater-element-i/next-greater-element-i.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
var nextGreaterElement = function(nums1, nums2) {
    const stack = []
    const hashmap = {}

    for (let num of nums2) {
        while (stack.length > 0 && num > stack[stack.length - 1]) {
            hashmap[stack.pop()] = num
        }
        stack.push(num)
    }

    while (stack.length > 0) {
        hashmap[stack.pop()] = -1
    }

    return nums1.map(num => hashmap[num] !== undefined ? hashmap[num] : -1)
};

File: 496-next-greater-element-i/next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        next_greater = {}
        stack = []

        for num in nums2:
            while stack and num > stack[-1]:
                smaller_num = stack.pop()
                next_greater[smaller_num] = num

            stack.append(num)

        return [next_greater.get(num, -1) for num in nums1]